runhouse/servers/env_servlet.py

```diff
@@ -177,5 +177,142 @@ async def adelete_local(self, key: Any):
     async def aclear_local(self):
         return await obj_store.aclear_local()
 
-    async def astatus_local(self, env_servlet_pid):
-        return obj_store.status_local(env_servlet_pid)
+    def get_node_ip(self):
+        import socket
+
+        return socket.gethostbyname(socket.gethostname())
+
+    def _get_env_cpu_usage(self):
+
+        import psutil
+
+        from runhouse.utils import get_pid
+
+        cluster_config = obj_store.cluster_config
+
+        total_memory = psutil.virtual_memory().total
+        node_ip = self.get_node_ip()
+        env_servlet_pid = get_pid()
+        node_name = ""  # will be set later. if not set here, an error will be raised.
+
+        if not cluster_config.get("resource_subtype") == "Cluster":
+            stable_internal_external_ips = cluster_config.get(
+                "stable_internal_external_ips"
+            )
+            for ips_set in stable_internal_external_ips:
+                internal_ip, external_ip = ips_set[0], ips_set[1]
+                if internal_ip == node_ip:
+                    # head ip == cluster address == cluster.ips[0]
+                    if ips_set[1] == cluster_config.get("ips")[0]:
+                        node_name = f"head ({external_ip})"
+                    else:
+                        node_name = f"worker_{stable_internal_external_ips.index(ips_set)} ({external_ip}"
+        else:
+            # a case it is a BYO cluster, assume that first ip in the ips list is the head.
+            ips = cluster_config.get("ips")
+            if len(ips) == 1 or node_ip == ips[0]:
+                node_name = f"head ({node_ip})"
+            else:
+                node_name = f"worker_{ips.index(node_ip)} ({node_ip})"
+
+        try:
+            env_servlet_process = psutil.Process(pid=env_servlet_pid)
+            memory_size_bytes = env_servlet_process.memory_full_info().uss
+            cpu_usage_percent = env_servlet_process.cpu_percent(interval=1)
+            env_memory_usage = {
+                "memory_size_bytes": memory_size_bytes,
+                "cpu_usage_percent": cpu_usage_percent,
+                "memory_percent_from_cluster": (memory_size_bytes / total_memory) * 100,
+                "total_cluster_memory": total_memory,
+                "env_memory_info": psutil.virtual_memory(),
+            }
+        except psutil.NoSuchProcess:
+            env_memory_usage = {}
+
+        return (env_memory_usage, node_name, total_memory, env_servlet_pid, node_ip)
+
+    def _get_env_gpu_usage(self, env_servlet_pid):
+        import subprocess
+
+        import runhouse as rh
+
+        # check it the cluster uses GPU or not
+        if rh.Package._detect_cuda_version_or_cpu() == "cpu":
```

Comment:
thoughts on whether we make this a separate utility as well? (instead of being part of the Package)

Reply:
yeah agreed, but i think maybe do it in a separate PR? Cause Package._detect_cuda_version_or_cpu() is being used in used places in the code. @jlewitt1